### PR TITLE
Third party dependencies contextual analysis for npm

### PR DESCRIPTION
**Pull request by EyalDelarea**
- [x] All [tests](https://github.com/jfrog/jfrog-cli-core#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-cli-core/actions/workflows/analysis.yml) passed.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

When enabling the new flag of the CLI :` third-party-contextual-analysis`
The applicability scanner will run on the environments folders as well. ( node_modules etc...)

In this case we will need to exclude the reported applicabilities which originated from the impacted source package.

For example: `protobufjs` CVE applicability was found:
applicability location is inside node_modules/protobufjs -> **disqualify** 
Other use case:
applicability location is inside node_modules/someOtherDep -> **valid**.

### CODE REVIEW

File: xray/commands/audit/jas/common.go [hunk 1]
```diff
@@ -19,8 +19,12 @@ import (
 	"gopkg.in/yaml.v3"
 )
 
+const (
+	NpmSkipPattern = "**/*node_modules*/**"
```
Comment (sverdlov93) [inline xray/commands/audit/jas/common.go:23]:
```suggestion
	nodeModulesPattern = "**/*node_modules*/**"
```

File: xray/utils/resultstable.go [hunk 2]
```diff
@@ -26,6 +26,10 @@ const (
 	rootIndex                  = 0
 	directDependencyIndex      = 1
 	directDependencyPathLength = 2
+
+	npmPackageTypeIdentifier = "npm://"
```
Comment (sverdlov93) [inline xray/utils/resultstable.go:30]: already exists

File: xray/utils/resultstable.go [hunk 3]
```diff
@@ -26,6 +26,10 @@ const (
...
+
+	nodeModules = "node_modules"
```
Comment (sverdlov93) [inline xray/utils/resultstable.go:32]: lets combine this with the pattern

File: xray/utils/resultstable.go [hunk 4]
```diff
@@ -89,10 +93,8 @@ func prepareViolations(violations []services.Violation, extendedResults *Extende
 		case "security":
 			cves := convertCves(violation.Cves)
 			applicableValue := getApplicableCveValue(extendedResults, cves)
-			if extendedResults.EntitledForJas {
-				for i := range cves {
-					cves[i].Applicability = getCveApplicability(cves[i], extendedResults.ApplicabilityScanResults)
-				}
+			for i, cve := range cves {
+				cves[i].Applicability = getCveApplicability(cve, extendedResults.ApplicabilityScanResults, nil)
```
Comment (sverdlov93) [inline xray/utils/resultstable.go:97]: you remove my if

File: xray/utils/resultstable.go [hunk 5]
```diff
@@ -211,10 +213,8 @@ func prepareVulnerabilities(vulnerabilities []services.Vulnerability, extendedRe
 		}
 		cves := convertCves(vulnerability.Cves)
 		applicableValue := getApplicableCveValue(extendedResults, cves)
-		if extendedResults.EntitledForJas {
-			for i := range cves {
-				cves[i].Applicability = getCveApplicability(cves[i], extendedResults.ApplicabilityScanResults)
-			}
+		for i, cve := range cves {
+			cves[i].Applicability = getCveApplicability(cve, extendedResults.ApplicabilityScanResults, vulnerability.Components)
```
Comment (sverdlov93) [inline xray/utils/resultstable.go:217]: here also

File: xray/utils/resultstable.go [hunk 6]
```diff
@@ -975,6 +975,10 @@ func getCveApplicability(cve formats.CveRow, applicabilityScanResults []*sarif.R
 
 		// Add new evidences from locations
 		for _, location := range foundResult.Locations {
+			fileName := GetLocationFileName(location)
```
Comment (sverdlov93) [inline xray/utils/resultstable.go:978]:
```suggestion
			fileName := GetRelativeLocationFileName(location, applicabilityRun.Invocations)
```

File: xray/utils/resultstable.go [hunk 7]
```diff
@@ -975,6 +975,10 @@ func getCveApplicability(cve formats.CveRow, applicabilityScanResults []*sarif.R
...
+			if shouldDisqualifyNpmEvidence(components, fileName) {
+				continue
+			}
 			applicability.Evidence = append(applicability.Evidence, formats.Evidence{
 				Location: formats.Location{
 					File:        GetRelativeLocationFileName(location, applicabilityRun.Invocations),
```
Comment (sverdlov93) [inline xray/utils/resultstable.go:984]:
```suggestion
					File:        fileName,
```

File: xray/utils/resultstable.go [hunk 8]
```diff
@@ -1002,3 +1011,42 @@ func printApplicableCveValue(applicableValue ApplicabilityStatus, isTable bool)
...
+// is inside the source code of the same package, we should disqualify it.
+//
+// For example,
+// Cve applicability was found inside the 'mquery' package.
+// filePath = myProject/node_modules/mquery/badCode.js , disqualify = True.
+// Disqualify the above evidence, as the reported applicability is used inside its own package.
+//
+// filePath = myProject/node_modules/mpath/badCode.js  , disqualify = False.
+// Found use of a badCode inside the node_modules from a different package, report applicable.
+func shouldDisqualifyNpmEvidence(components map[string]services.Component, evidenceFilePath string) (disqualify bool) {
```
Comment (sverdlov93) [inline xray/utils/resultstable.go:1027]:
```suggestion
func shouldDisqualifyEvidence(components map[string]services.Component, evidenceFilePath string) (disqualify bool) {
```

File: xray/utils/resultstable.go [hunk 9]
```diff
@@ -1002,3 +1011,42 @@ func printApplicableCveValue(applicableValue ApplicabilityStatus, isTable bool)
...
+	for key := range components {
+		dependencyName := extractNpmDependencyNameFromComponent(key)
+		if dependencyName == "" {
+			return
+		}
+		if strings.Contains(evidenceFilePath, nodeModules+"/"+dependencyName) {
+			disqualify = true
+			return
```
Comment (sverdlov93) [inline xray/utils/resultstable.go:1035]:
```suggestion
			return true
```

File: xray/utils/resultstable.go (hunk 3 above)
Comment (EyalDelarea) [reply xray/utils/resultstable.go]: It is used separately

File: xray/commands/audit/jas/common.go [hunk 10]
```diff
@@ -19,8 +19,12 @@ import (
 	"gopkg.in/yaml.v3"
 )
 
+const (
+	NodeModulesPattern = "**/*node_modules*/**"
```
Comment (sverdlov93) [inline xray/commands/audit/jas/common.go:23]:
```suggestion
	nodeModulesPattern = "**/*node_modules*/**"
```

File: xray/commands/audit/scarunner.go [hunk 11]
```diff
@@ -115,6 +114,18 @@ func runScaScanOnWorkingDir(params *AuditParams, results *Results, workingDir, r
...
+// When building pip dependency tree using pipdeptree, some of the direct dependencies are recognized as transitive and missed by the CA scanner.
+// Our solution for this case is to send all dependencies to the CA scanner.
+// When thirdPartyApplicabilityScan is true, use flatten graph to include all the dependencies in applicability scanning.
+// Only npm is supported for this flag.
+func shouldUseAllDependencies(thirdPartyApplicabilityScan bool, tech coreutils.Technology) bool {
+	if tech == coreutils.Pip || (thirdPartyApplicabilityScan && tech == coreutils.Npm) {
+		return true
+	} else {
+		return false
+	}
```
Comment (sverdlov93) [inline xray/commands/audit/scarunner.go:126]:
```suggestion
	}
	return false

```

Comment (sverdlov93) [inline xray/commands/audit/scarunner.go:126]:
```suggestion
	return tech == coreutils.Pip || (thirdPartyApplicabilityScan && tech == coreutils.Npm)
```

File: xray/commands/audit/jas/common.go (hunk 10 above)
Comment (EyalDelarea) [reply xray/commands/audit/jas/common.go:23]: been used from applicablitymanager.go so has to be public

**Pull request merged**